semgrep/semgrep/output.py

```diff
@@ -173,3 +187,178 @@ def build_output(
+        sys.exit(exit_code)
+
+
+class OutputHandler:
+    """
+    Handle all output in a central location. Rather than calling `print_error` directly,
+    you should call `handle_*` as appropriate.
+
+    In normal usage, it should be constructed via the contextmanager, `managed_output`. It ensures that everything
+    is handled properly if exceptions are thrown.
+
+    If you need to stop execution immediately (think carefully if you really want this!), throw an exception.
+    If this is normal behavior, the exception _must_ inherit from `SemgrepError`.
+
+    If you want execution to continue, _report_ the exception via the appropriate `handle_*` method.
+    """
+
+    def __init__(
+        self,
+        output_settings: OutputSettings,
+        stderr: IO = sys.stderr,
+        stdout: IO = sys.stdout,
+    ):
+        self.settings = output_settings
+        self.stderr = stderr
+        self.stdout = stdout
+
+        self.rule_matches: List[RuleMatch] = []
+        self.debug_steps_by_rule: Dict[Rule, List[Dict[str, Any]]] = {}
+        self.rules: FrozenSet[Rule] = frozenset()
+        self.semgrep_core_errors: List[Dict[str, Any]] = []
+        self.semgrep_rule_errors: List[SemgrepError] = []
+        self.has_output = False
+
+        self.exit_code = 0
+        self.final_error: Optional[Exception] = None
+
+    def handle_semgrep_core_errors(self, semgrep_errors: List[Dict[str, Any]]) -> None:
+        """
+        Report errors coming directly from semgrep-core (raw JSON objects)
+        """
+        self.semgrep_core_errors += semgrep_errors
+        if self.settings.output_format == OutputFormat.TEXT:
+            for error in semgrep_errors:
+                print_error(pretty_error(error))
+
+    def handle_semgrep_rule_errors(self, error: SemgrepError) -> None:
+        """
+        Report parse errors from semgrep rules. Either:
+        - when the pattern fails to parse in semgrep-core
+        - when the YAML or YAML structure is invalid
+        """
+        self.semgrep_rule_errors.append(error)
+        self._output_exception(error)
+
+    def handle_semgrep_core_output(
+        self,
+        rule_matches_by_rule: Dict[Rule, List[RuleMatch]],
+        debug_steps_by_rule: Dict[Rule, List[Dict[str, Any]]],
+    ) -> None:
+        self.has_output = True
+        self.rules = self.rules.union(rule_matches_by_rule.keys())
+        self.rule_matches += [
+            match
+            for matches_of_one_rule in rule_matches_by_rule.values()
+            for match in matches_of_one_rule
+        ]
+
+        self.debug_steps_by_rule.update(debug_steps_by_rule)
+
+    def handle_unhandled_exception(self, ex: Exception) -> None:
+        """
+        This is called by the context manager upon an unhandled exception. If you want to record a final
+        error & set the exit code, but keep executing to perform cleanup tasks, call this method.
+        """
+        if isinstance(ex, SemgrepError):
```

Comment:
This feels kinda hamfisted compared to using Python's builtin mechanism for doing this:

```python
try:
    ...
except ExceptionType1:
    ...
except ExceptionType2:
    ...
```

Reply:
I sort of agree, I'm abusing exceptions are more general purpose errors here.

We want to be able to merely record that an error occurred and keep going, hence passing exception objects around rather than throwing them. This is replacing code, that simply called `print_error("some string")`.

But to be able to incorporate output for all stages of the pipeline into JSON, we need a more structured approach.